tests/src/test/scala/whisk/core/containerpool/docker/test/AkkaContainerClientTests.scala

```diff
@@ -0,0 +1,213 @@
+  implicit val ec = actorSystem.dispatcher
+
+  var testHang: FiniteDuration = 0.second
+  var testStatusCode: Int = 200
+  var testResponse: String = null
+  var testConnectionFailCount: Int = 0
+
+  val mockServer = new LocalServerTestBase {
+    var failcount = 0
+    override def setUp() = {
+      super.setUp()
+      this.serverBootstrap
+        .registerHandler(
+          "/init",
+          new HttpRequestHandler() {
+            override def handle(request: HttpRequest, response: HttpResponse, context: HttpContext) = {
+              if (testHang.length > 0) {
+                Thread.sleep(testHang.toMillis)
+              }
+              if (testConnectionFailCount > 0 && failcount < testConnectionFailCount) {
+                failcount += 1
+                println("failing in test")
+                throw new RuntimeException("failing...")
+              }
+              response.setStatusCode(testStatusCode);
+              if (testResponse != null) {
+                response.setEntity(new StringEntity(testResponse, StandardCharsets.UTF_8))
+              }
+            }
+          })
+    }
+  }
+
+  mockServer.setUp()
+  val httpHost = mockServer.start()
+  val hostWithPort = s"${httpHost.getHostName}:${httpHost.getPort}"
+
+  before {
+    testHang = 0.second
+    testStatusCode = 200
+    testResponse = null
+    testConnectionFailCount = 0
+    stream.reset()
+  }
+
+  override def afterAll = {
+    mockServer.shutDown()
+  }
+
+  behavior of "PoolingContainerClient"
+
+  it should "not wait longer than set timeout" in {
+    val timeout = 5.seconds
+    val connection = new AkkaContainerClient(httpHost.getHostName, httpHost.getPort, timeout, 1.B, 100)
+    testHang = timeout * 2
+    val start = Instant.now()
+    val result = Await.result(connection.post("/init", JsObject.empty, retry = true), 10.seconds)
+
+    val end = Instant.now()
+    val waited = end.toEpochMilli - start.toEpochMilli
+    result shouldBe 'left
+    waited should be > timeout.toMillis
+    waited should be < (timeout * 2).toMillis
+  }
+
+  it should "handle empty entity response" in {
+    val timeout = 5.seconds
+    val connection = new AkkaContainerClient(httpHost.getHostName, httpHost.getPort, timeout, 1.B, 100)
+    testStatusCode = 204
+    val result = Await.result(connection.post("/init", JsObject.empty, retry = true), 10.seconds)
+    result shouldBe Left(NoResponseReceived())
+  }
+
+  it should "retry till timeout on StreamTcpException" in {
+    val timeout = 5.seconds
+    val connection = new AkkaContainerClient("0.0.0.0", 12345, timeout, 1.B, 100)
+    val start = Instant.now()
+    val result = Await.result(connection.post("/init", JsObject.empty, retry = true), 10.seconds)
+    val end = Instant.now()
+    val waited = end.toEpochMilli - start.toEpochMilli
+    result should be('left)
+    result.left.get shouldBe a[Timeout]
+    result.left.get.asInstanceOf[Timeout].t shouldBe a[TimeoutException]
```

Comment:
Same as below, could be rewritten to:

```scala
result match {
  case Left(Timeout(_ : TimeoutException) => // good
  case _ => fail(...)
}
```

IMO that pronounces the nesting of the exceptions more and is clearer in what you're actually testing. WDYT?

Reply:
agreed